src/server.js

```diff
@@ -0,0 +1,96 @@
+'use strict';
+
+const moment = require('moment');
+const cudl = require('cuddle');
+const _ = require('lodash');
+
+const config = require('./config');
+const cache = require('./cache');
+
+module.exports = {
+    verify_scopes
+};
+
+function verify_scopes (required_scopes) {
+    return (req, res, next) => {
+        return check_scopes(req, required_scopes)
+            .then(result => {
+                req.client_id = result.client_id;
+                next();
+            })
+            .catch(next);
+    };
+}
+
+function check_scopes (request, required_scopes) {
+    const access_token = request.headers['access-token'];
+
+    return check_input(access_token)
+        .then(get_token_info)
+        .then(result => ({access_token, result, required_scopes}))
+        .then(check_token_validity);
+}
+
+function check_input (access_token) {
+    return new Promise((resolve, reject) => {
+        if (!access_token) {
+            return reject(new Error('Access token is required.'));
+        }
+
+        resolve(access_token);
+    });
+}
+
+function get_token_info (access_token) {
+    const cached_token_info_result = cache.get('server', access_token);
+
+    if (_.has(cached_token_info_result, 'date_expiration')) {
+        const date_expiration = moment.utc(cached_token_info_result.date_expiration, 'x');
+        const token_is_expired = date_expiration.isSameOrBefore(moment());
+
+        if (!token_is_expired) {
+            return Promise.resolve(cached_token_info_result);
+        }
+
+        cache.forget('server', access_token);
+    }
+
+    return cudl.get
+        .to(config.base_url + config.path + '/oauth/tokeninfo')
+        .send({access_token: access_token})
+        .promise()
+        .then(result => {
+            if (result.scopes) {
+                cache.set('server', access_token, result, config.server_expiry);
+            }
+
+            return result;
+        })
+        .catch(error => Promise.reject(
+            _.has(error, 'response.message')
+                ? error.response
+                : error
+            )
+        );
+
+}
+
+function check_token_validity ({result, access_token, required_scopes}) {
+    return new Promise((resolve, reject) => {
+        if (!result.scopes) {
+            cache.forget('server', access_token);
+            return reject(new Error('Something went wrong, server did not return scopes, please try again.'));
+        }
+
+        const result_scopes = result.scopes.split(' ');
+        const has_valid_scopes = _.intersection(result_scopes, required_scopes).length;
+
+        if (has_valid_scopes) {
```

Comment:
Should we also check for date_expiration again since /oauth/tokeninfo could tell us that the access_token has been revoked?

Reply:
What do you mean? Freedom Accounts responds with a non-200 token expired error when the token is expired.

Reply:
I mean, shouldn't we be removing whatever was cached when there's an error for get_token_info?
ex L80, L69

Reply:
get_token_info is not cached when there is an error so there is no need to delete from cache (i.e. in L69).  In the case of failing the has_valid_scopes, we still want to cache whatever scope the access token actually had in case it is used again in another endpoint where it has valid scopes.

Reply:
I agree with the has_valid_scopes check part.
However, if get_token_info fails, doesn't it mean the access token is already invalid and we should be requesting a new access_token?

Reply:
get_token_info is only called if there is no cached token info for the specified access token so if it fails, the next call will still do the request to tokeninfo.

Reply:
Yes, but it would also likely to fail since it'll be for the same access_token, right? So shouldn't we actually force to get a new access_token instead?

Reply:
Request to get new access token? The requesting of a new access token happens on the client side.

* If the token is still cached, then tokeninfo will not be called and the info in the cache will be used.
* When there's a call and the cache is expired (in 5 minutes and it automatically gets deleted in cache too), the token may still be valid (token is for 1 hour) so tokeninfo is called again to update.
* If there is no error or issues, the new result is cached.
* Otherwise, the call is rejected and token cache remains deleted (which will make it request token info the next time).

Reply:
Right, confused with other parts of the code. Thought we need to get the access_token from accounts, but access_token is passed in.